source/pidomus_checkpoint_restart.cc

```diff
@@ -0,0 +1,411 @@
+    std::ostringstream oss;
+
+    // serialize into a stringstream
+    boost::archive::binary_oarchive oa (oss);
+    oa << (*this);
+
+    // compress with zlib and write to file on the root processor
+#ifdef DEAL_II_WITH_ZLIB
+    if (my_id == 0)
+      {
+        uLongf compressed_data_length = compressBound (oss.str().length());
+        std::vector<char *> compressed_data (compressed_data_length);
+        int err = compress2 ((Bytef *) &compressed_data[0],
+                             &compressed_data_length,
+                             (const Bytef *) oss.str().data(),
+                             oss.str().length(),
+                             Z_BEST_COMPRESSION);
+        (void)err;
+        Assert (err == Z_OK, ExcInternalError());
+
+        // build compression header
+        const uint32_t compression_header[4]
+          = { 1,                                   /* number of blocks */
+              (uint32_t)oss.str().length(), /* size of block */
+              (uint32_t)oss.str().length(), /* size of last block */
+              (uint32_t)compressed_data_length
+            }; /* list of compressed sizes of blocks */
+
+        std::ofstream f ((snap_prefix + "restart.resume.z").c_str());
+        f.write((const char *)compression_header, 4 * sizeof(compression_header[0]));
+        f.write((char *)&compressed_data[0], compressed_data_length);
+      }
+#else
+    AssertThrow (false,
+                 ExcMessage ("You need to have deal.II configured with the 'libz' "
+                             "option to support checkpoint/restart, but deal.II "
+                             "did not detect its presence when you called 'cmake'."));
+#endif
+
+  }
+  pcout << "*** Snapshot created!" << std::endl << std::endl;
+}
+
+
+template <int dim, int spacedim, typename LAC>
+void
+piDoMUS<dim,spacedim,LAC>::
+save_solutions_and_triangulation(const LADealII::VectorType &y,
+                                 const LADealII::VectorType &y_dot,
+                                 const LADealII::VectorType &y_expl,
+                                 const LADealII::VectorType &,
+                                 const LADealII::VectorType &) const
+{
+  std::string sy = snap_prefix+"y.bin";
+  std::string sdot = snap_prefix+"y_dot.bin";
+  std::string sexpl = snap_prefix+"y_expl.bin";
+  const char *file_y = sy.c_str();
+  const char *file_y_dot = sdot.c_str();
+  const char *file_y_expl = sexpl.c_str();
+
+  if (file_exists(file_y))
+    copy_file(file_y,snap_prefix+"y.bin.old");
+  if (file_exists(file_y_dot))
+    copy_file(file_y_dot,snap_prefix+"y_dot.bin.old");
+  if (file_exists(file_y_expl))
+    copy_file(file_y_expl,snap_prefix+"y_expl.bin.old");
+
+  std::ofstream out_y (file_y);
+  std::ofstream out_y_dot (file_y_dot);
+  std::ofstream out_y_expl (file_y_expl);
+
+  y.block_write(out_y);
+  out_y.close();
+
+  y_dot.block_write(out_y_dot);
+  out_y_dot.close();
+
+  y_expl.block_write(out_y_expl);
+  out_y_expl.close();
+
```

Comment:
and at resume time:

~~~

  eigenfunctions.resize (n_eigenvalues);
  eigenfunctions_ghosted.resize (n_eigenvalues);

  std::vector<PETScWrappers::MPI::Vector *> all_out(n_eigenvalues);
  for (unsigned int i=0; i<n_eigenvalues; ++i)
    {
      eigenfunctions[i].reinit (owned_index_set, MPI_COMM_WORLD);
      all_out[i] = &eigenfunctions[i];
    }

  // Ghosted functions are used internally by soltrans
  if (cycle > 0 && resume_computation == false)
    soltrans->interpolate(all_out);
  else if (resume_computation == true)
    {
      soltrans = SP(new parallel::distributed::SolutionTransfer<dim, PETScWrappers::MPI::Vector>(*dof_handler));
      soltrans->deserialize(all_out);
      resume_computation = false;
    }


  // Now that we have transfered the old ones, reinitialize ghosted vectors.
  for (unsigned int i=0; i<n_eigenvalues; ++i)
    {
      eigenfunctions_ghosted[i].reinit (owned_index_set,
                                        relevant_index_set,
                                        MPI_COMM_WORLD);
      eigenfunctions_ghosted[i] = eigenfunctions[i];
    }
~~~